### PR TITLE
Refactor fetcher, polling strategy and manager

### PR DESCRIPTION
**Pull request by mariokostelac**
Simplifying manager, fetcher and polling strategy in order to make whole system easier to understand, test and more performant
- Fetcher became synchronous and quite simple. It became totally unaware of the manager so we finally lost that bidirectional dependency.
- Manager became far slimmer. It lost responsibility of deciding on next queue (moved to polling strategy) and calls fetcher and assign method synchronously for the sake of easier testing.
- Polling strategy it quite simple and can be replaced with something else. WeightedRoundRobin actually implements the old implicit strategy.

I've done some testing locally for next setup
- one 1, two workers
- workers just re-enq the job back to the q
- run OS X 10.11.6, MacBook Pro (Retina, 13-inch, Early 2015)
- run workers for 5 minutes

For concurrency set to 1, throughput increased ~50%.
For concurrency set to 10, throughput increased ~30%.
For concurrency set to 20, throughput increased ~20%.

It's important to note that increase becomes higher if jobs gets faster (shorter).

WDYT @phstc?

### CODE REVIEW

File: lib/shoryuken/fetcher.rb [hunk 1]
```diff
@@ -1,73 +1,44 @@
...
+          limit = available_processors > FETCH_LIMIT ? FETCH_LIMIT : available_processors
 
+          sqs_msgs = Array(receive_messages(queue, limit))
+          logger.info { "Found #{sqs_msgs.size} messages for '#{queue}'" }
           logger.debug { "Fetcher for '#{queue}' completed in #{elapsed(started_at)} ms" }
+          sqs_msgs
         rescue => ex
           logger.error { "Error fetching message: #{ex}" }
           logger.error { ex.backtrace.first }
+          []
```
Comment (phstc) [inline lib/shoryuken/fetcher.rb]: @mariokostelac not sure if we need `rescue => ex` where, as we are using [`watchdog`](https://github.com/phstc/shoryuken/blob/bf65b6e90bd6648279c89eeaafeb3f21cbbfd1c1/lib/shoryuken/util.rb#L3) wrapping the method body.

File: lib/shoryuken/fetcher.rb [hunk 2]
```diff
@@ -1,73 +1,44 @@
...
 
-    def patch_sqs_msgs!(sqs_msgs)
-      sqs_msgs.instance_eval do
-        def message_id
-          "batch-with-#{size}-messages"
-        end
-      end
+    def receive_messages(queue, limit)
+      # AWS limits the batch size by 10
+      limit = limit > FETCH_LIMIT ? FETCH_LIMIT : limit
```
Comment (phstc) [inline lib/shoryuken/fetcher.rb]: @mariokostelac we are doing this same `FETCH_LIMIT` checking in the `fetch` method too, I believe we should do it only on the `receive_messages` method, as it's the one responsible for receiving messages.

File: lib/shoryuken/manager.rb [hunk 3]
```diff
@@ -154,61 +128,41 @@ def real_thread(proxy_id, thr)
...
-          dispatch
-        end
+        processor.async.process(queue, sqs_msg)
       end
     end
 
-    def current_queue_weight(queue)
-      queue_weight(@queues, queue)
+    def dispatch_batch(queue)
+      batch = fetcher.fetch(queue, BATCH_LIMIT)
```
Comment (phstc) [inline lib/shoryuken/manager.rb]: @mariokostelac I believe we could use something like `Fetcher::FETCH_LIMIT` instead of `BATCH_LIMIT` as it should be the same, in case of batch we want the maximum possible to fetch anyway == fetch limit, and the fetch is the one that knows the what's the maximum.

Comment (phstc) [inline lib/shoryuken/manager.rb]: To reduce the dependency with constants, we could also have specific methods `fetch_max` and `fetch_limited` or something like that, but I don't see problems with `Fetcher::FETCH_LIMIT` as `Manager` already knows `Fetcher`.

File: lib/shoryuken/manager.rb [hunk 4]
```diff
@@ -241,5 +195,15 @@ def hard_shutdown_in(delay)
         end
       end
     end
+
+    def patch_batch!(sqs_msgs)
+      sqs_msgs.instance_eval do
+        def message_id
+          "batch-with-#{size}-messages"
+        end
```
Comment (phstc) [inline lib/shoryuken/manager.rb]: @mariokostelac why is this needed?

**Comment by phstc** [comment]
Hey @mariokostelac it's looking super good. I did an initial PR code review, but I'm going to checkout it and test it locally, so I can understand it better.

File: lib/shoryuken/fetcher.rb (hunk 1 above)
Comment (mariokostelac) [inline lib/shoryuken/fetcher.rb]: We should use one or another, not both, that's for sure. I do like having error handling here because SQS does return 500 now and then and it'd be nice to return empty array / retry fetch in that case.

File: lib/shoryuken/manager.rb (hunk 3 above)
Comment (mariokostelac) [inline lib/shoryuken/manager.rb]: It was logical separation, but I am good with referencing fetch limit. Other approach would be having `BATCH_LIMIT = Fetcher::FETCH_LIMIT`

File: lib/shoryuken/manager.rb (hunk 4 above)
Comment (mariokostelac) [inline lib/shoryuken/manager.rb]: i have to check it out, it was part of fetcher before and i just moved it here. I'd like to get rid of that if possible. Will check if we need this (it think we did need it for logging).

File: lib/shoryuken/manager.rb (hunk 3 above)
Comment (phstc) [inline lib/shoryuken/manager.rb]: @mariokostelac I wouldn't introduce `BATCH_LIMIT`, because we don't have a such thing. In this case we want: give the maximum you can. Just to make it simpler for future maintenance.

File: lib/shoryuken/fetcher.rb (hunk 1 above)
Comment (phstc) [inline lib/shoryuken/fetcher.rb]: @mariokostelac hm nvm just saw that the [current implementation](https://github.com/phstc/shoryuken/blob/master/lib/shoryuken/fetcher.rb#L51-L54)  does the same.

**Comment by phstc** [comment]
@mariokostelac I could perform an end-to-end test today. It seems to be working fine, but `ctrl-c` does not seem to be working, it seems it isn't catching the shutdown signal, besides that and two feedbacks I left to you, I believe it's ready to go.

**Comment by mariokostelac** [comment]
@phstc ok, I will apply changes based on your comments and take a look at Ctrl-C issue. It worked for me (that's how I stopped it), but could be I broke something.

**Comment by phstc** [comment]
@mariokostelac if `Ctrl-C` (`SIGINT`) works fine for you, ping me and I can test it again. I don't see anything related to trapping signals in your PR, might be something on my test, but it's good to double check it.

**Comment by atyndall** [comment]
Has any progress been made on this? The `StrictPriorityPolling` implementation in pull #106 would be ideal for our circumstances.

**Comment by phstc** [comment]
@mariokostelac let's merge it :shipit: 🍻 so, we can also merge #263. Could you check the merge the conflicts?

**Comment by mariokostelac** [comment]
@phstc I see I have not write rights anymore. Do you prefer creating another PR for that (from forked repo)?

**Comment by phstc** [comment]
@mariokostelac I think that makes sense.

**Comment by mariokostelac** [comment]
Closing in favor of https://github.com/phstc/shoryuken/pull/236!
